### PR TITLE
[Navigation Material] Update Navigation Material after ModalBottomSheetLayout bug fixes

### PR DESCRIPTION
**Pull request by jossiwolf**
This PR refactors our Navigation Material code after upstream updates to ModalBottomSheetLayout that address one-frame consistency and anchor update + reconciliation logic.

We now retain the back stack entry until all animations are completed.

### API Changes
|API|Status|
|---|---|
|public ctor BottomSheetNavigatorSheetState|internal ctor|
|BottomSheetNavigatorSheetState#offset|Deprecated|
|BottomSheetNavigatorSheetState#direction|Deprecated|
|BottomSheetNavigatorSheetState#progress|Deprecated|

### Other changes
- BottomSheetNavigator now uses moveable content under the hood
- The strategy for retaining bottom sheet back stack entries has changed to only dispose of them after the sheet has been hidden

Fixes: #772, #813, #910, #937, #1031, #1147, #1356

### CODE REVIEW

File: navigation-material/src/main/java/com/google/accompanist/navigation/material/BottomSheetNavigator.kt [hunk 1]
```diff
@@ -165,63 +177,78 @@ public class BottomSheetNavigator(
...
-        DisposableEffect(backStackEntries) {
-            transitionsInProgressEntries.forEach {
-                if (it != latestEntry) state.markTransitionComplete(it)
+            val retainedEntry by produceState<NavBackStackEntry?>(initialValue = null) {
+                backStack
+                    .map { backStackEntries ->
+                        backStackEntries.lastOrNull { entry ->
+                            // We might have entries in the back stack that aren't started
+                            // currently, so filter these
+                            entry.lifecycle.currentState.isAtLeast(Lifecycle.State.INITIALIZED)
```
Comment (jossiwolf) [inline navigation-material/src/main/java/com/google/accompanist/navigation/material/BottomSheetNavigator.kt:197]: @jbw0033 in tests (specifically, `testTapOnScrimDismissesSheetAndPopsBackStack`) the entry added would only be `INITIALIZED`. All is fine in the sample. Any ideas what I'm missing? Do we even need to filter?

**Comment by jossiwolf** [comment]
@jbw0033 `testSheetShownAfterNavControllerRestoresState` also seems to fail now - looks like the sheet state doesn't get moved to Hidden. Do you have an idea?

File: sample/src/main/java/com/google/accompanist/sample/navigation/material/BottomSheetNavSample.kt [hunk 2]
```diff
@@ -101,7 +116,17 @@ private fun HomeScreen(showSheet: () -> Unit, showFeed: () -> Unit) {
 
 @Composable
 private fun BottomSheet(showFeed: () -> Unit, showAnotherSheet: () -> Unit, arg: String) {
+    var items by remember { mutableStateOf(2) }
```
Comment (jossiwolf) [inline sample/src/main/java/com/google/accompanist/sample/navigation/material/BottomSheetNavSample.kt:119]: Will undo the sample changes before merging :)

Comment (jbw0033) [reply sample/src/main/java/com/google/accompanist/sample/navigation/material/BottomSheetNavSample.kt:119]: ACK

Comment (jbw0033) [reply sample/src/main/java/com/google/accompanist/sample/navigation/material/BottomSheetNavSample.kt:119]: Is it worth having multiple samples to test specific cases?

File: navigation-material/src/main/java/com/google/accompanist/navigation/material/BottomSheetNavigator.kt [hunk 3]
```diff
@@ -165,63 +177,78 @@ public class BottomSheetNavigator(
...
-                // Sheet dismissal can be started through popBackStack in which case we have a
-                // transition that we'll want to complete
-                if (transitionsInProgressEntries.contains(backStackEntry)) {
-                    state.markTransitionComplete(backStackEntry)
-                } else {
-                    state.pop(popUpTo = backStackEntry, saveState = false)
+            // Mark all of the entries' transitions as complete, except for the entry we are
+            // currently displaying because it will have its transition completed when the sheet's
+            // animation has completed
+            DisposableEffect(transitionsInProgressEntries) {
```
Comment (jbw0033) [inline navigation-material/src/main/java/com/google/accompanist/navigation/material/BottomSheetNavigator.kt:211]: Can we still change the system so that we always do the show/hide so that we get callbacks in all cases. And then the `onSheetShown` and `onSheetDismissed` callbacks can just do:

```
transitionsInProgressEntries.forEach {
    state.markTransitionComplete(it)
}
```

And this big of code should no longer be needed I believe.

Comment (jossiwolf) [reply navigation-material/src/main/java/com/google/accompanist/navigation/material/BottomSheetNavigator.kt]: Done

File: gradle/libs.versions.toml [hunk 4]
```diff
@@ -1,9 +1,9 @@
 [versions]
 
-compose = "1.3.1"
+compose = "1.4.0-SNAPSHOT"
```
Comment (jossiwolf) [inline gradle/libs.versions.toml:3]: Will revert this change after the next release before merging

File: sample/src/main/java/com/google/accompanist/sample/navigation/material/BottomSheetNavSample.kt (hunk 2 above)
Comment (jossiwolf) [reply sample/src/main/java/com/google/accompanist/sample/navigation/material/BottomSheetNavSample.kt:119]: Not specifically samples I think, but I expanded tests to cover this (`testSheetContentSizeChangeDuringAnimation_opensSheet_shortSheetToShortSheet`, `testSheetContentSizeChangeDuringAnimation_opensSheet_shortSheetToTallSheet`, `testSheetContentSizeChangeDuringAnimation_opensSheet_tallSheetToTallSheet`, `testSheetContentSizeChangeDuringAnimation_opensSheet_tallSheetToShortSheet`)

File: navigation-material/src/androidTest/java/com/google/accompanist/navigation.material/BottomSheetNavigatorTest.kt [hunk 5]
```diff
@@ -273,9 +296,6 @@ internal class BottomSheetNavigatorTest {
 
         assertThat(navigatorState.transitionsInProgress.value)
             .doesNotContain(backStackEntry1)
-
-        assertThat(navigatorState.transitionsInProgress.value)
```
Comment (ianhanniballake) [inline navigation-material/src/androidTest/java/com/google/accompanist/navigation.material/BottomSheetNavigatorTest.kt:277]: Is there a reason this was removed? I would think we still want to ensure after `awaitIdle` completes that `backStackEntry2` has been removed from the transitioning set and moved all the way to `RESUMED`.

File: navigation-material/src/androidTest/java/com/google/accompanist/navigation.material/BottomSheetNavigatorTest.kt [hunk 6]
```diff
@@ -368,6 +388,379 @@ internal class BottomSheetNavigatorTest {
...
+        val animationDuration = 2000
+        val animationSpec = tween<Float>(animationDuration)
+        lateinit var navigator: BottomSheetNavigator
+        lateinit var navController: NavHostController
+        var height: Dp by mutableStateOf(20.dp)
+
+        composeTestRule.setContent {
+            navigator = rememberBottomSheetNavigator(animationSpec)
+            navController = rememberNavController(navigator)
+            navController.currentBackStackEntryAsState().value
```
Comment (ianhanniballake) [inline navigation-material/src/androidTest/java/com/google/accompanist/navigation.material/BottomSheetNavigatorTest.kt:402]: Why do we need to read the value here? This should work whether or not we add extra recompositions, right?

File: navigation-material/src/main/java/com/google/accompanist/navigation/material/BottomSheetNavigator.kt [hunk 7]
```diff
@@ -132,9 +141,9 @@ public fun rememberBottomSheetNavigator(
 @ExperimentalMaterialNavigationApi
 @OptIn(ExperimentalMaterialApi::class)
 @Navigator.Name("BottomSheetNavigator")
-public class BottomSheetNavigator(
+class BottomSheetNavigator internal constructor(
```
Comment (ianhanniballake) [inline navigation-material/src/main/java/com/google/accompanist/navigation/material/BottomSheetNavigator.kt:144]: Can you explain why this is in an `internal` constructor now? If this is intentional, make sure to add a comment stating that should only be constructed via `rememberBottomSheetNavigator` to the KDocs if that is indeed the only entry point supported.

File: navigation-material/src/main/java/com/google/accompanist/navigation/material/BottomSheetNavigator.kt [hunk 8]
```diff
@@ -165,63 +174,71 @@ public class BottomSheetNavigator(
...
             }
-        )
-    }
+
+            SheetContentHost(
+                backStackEntry = retainedEntry,
+                sheetState = sheetState,
+                saveableStateHolder = saveableStateHolder,
+                onSheetShown = {
+                    transitionsInProgressEntries.forEach(state::markTransitionComplete)
```
Comment (ianhanniballake) [inline navigation-material/src/main/java/com/google/accompanist/navigation/material/BottomSheetNavigator.kt:214]: How does this work when moving between multiple screens (i.e., from Sheet A then `navigate(SheetB)`, `navigate(SheetC)` - do we move directly to `SheetC` (in which case this makes sense - finishing the show of `SheetC` should mark both `SheetB` and `SheetC` as completed) or do we see `SheetB`, then subsequently see `SheetC`?

File: navigation-material/src/androidTest/java/com/google/accompanist/navigation.material/BottomSheetNavigatorTest.kt [hunk 9]
```diff
@@ -368,6 +388,379 @@ internal class BottomSheetNavigatorTest {
...
+        }
+
+        composeTestRule.mainClock.autoAdvance = true
+
+        composeTestRule.waitForIdle()
+        composeTestRule.runOnUiThread {
+            navController.navigate("sheet")
+        }
+        composeTestRule.mainClock.advanceTimeBy(100)
+        height = (composeTestRule.onRoot().getUnclippedBoundsInRoot().height) / 3
```
Comment (ianhanniballake) [inline navigation-material/src/androidTest/java/com/google/accompanist/navigation.material/BottomSheetNavigatorTest.kt:422]: Can we verify that the sheet is in `transitionsInProgress` here while the sheet is being opened?

Similarly, it would be good to verify that it has been removed from that set after the animation finishes, here and in the other tests as well.

File: navigation-material/src/androidTest/java/com/google/accompanist/navigation.material/BottomSheetNavigatorTest.kt (hunk 6 above)
Comment (jossiwolf) [reply navigation-material/src/androidTest/java/com/google/accompanist/navigation.material/BottomSheetNavigatorTest.kt:402]: Yep, good catch!

File: navigation-material/src/androidTest/java/com/google/accompanist/navigation.material/BottomSheetNavigatorTest.kt (hunk 9 above)
Comment (jossiwolf) [reply navigation-material/src/androidTest/java/com/google/accompanist/navigation.material/BottomSheetNavigatorTest.kt:422]: Done

File: navigation-material/src/androidTest/java/com/google/accompanist/navigation.material/BottomSheetNavigatorTest.kt (hunk 6 above)
Comment (jossiwolf) [reply navigation-material/src/androidTest/java/com/google/accompanist/navigation.material/BottomSheetNavigatorTest.kt:402]: Yep, not needed. Done

File: navigation-material/src/main/java/com/google/accompanist/navigation/material/BottomSheetNavigator.kt (hunk 7 above)
Comment (jossiwolf) [reply navigation-material/src/main/java/com/google/accompanist/navigation/material/BottomSheetNavigator.kt:144]: Good call, will point it out and softly deprecate it for now. Folks are constructing navigators with custom sheet states which is not something we want to support, and `skipHalfExpanded` support will likely require changing the constructor so I'd like to not have to maintain binary compatibility for this in the future.

File: navigation-material/src/androidTest/java/com/google/accompanist/navigation.material/BottomSheetNavigatorTest.kt (hunk 5 above)
Comment (jossiwolf) [reply navigation-material/src/androidTest/java/com/google/accompanist/navigation.material/BottomSheetNavigatorTest.kt:277]: Ack, not sure why I removed it.

**Review by ianhanniballake — APPROVED** [review]

File: navigation-material/src/main/java/com/google/accompanist/navigation/material/BottomSheetNavigator.kt (hunk 8 above)
Comment (jossiwolf) [reply navigation-material/src/main/java/com/google/accompanist/navigation/material/BottomSheetNavigator.kt:227]: If we're being consistent with our UX we'd probably want to first see `SheetB` and then `SheetC`. I'm going to leave addressing this for a follow-up PR, I believe this behavior is okay for now.

File: navigation-material/src/main/java/com/google/accompanist/navigation/material/BottomSheetNavigator.kt (hunk 7 above)
Comment (jossiwolf) [reply navigation-material/src/main/java/com/google/accompanist/navigation/material/BottomSheetNavigator.kt]: Decided not to deprecate it for the time being. I saw some folks pass a sheet state with `skipHalfExpanded = true` which is a valid use case imo

**Review by jbw0033 — APPROVED** [review]
